mindmeld/components/entity_resolver.py

```diff
@@ -562,38 +1151,575 @@ def _predict_exact_match(self, entity):
+                            values.append(item_value)
+                values_list.append(values)
+        except (KeyError, TypeError) as e:
+            logger.error(e)
+            return None
+
+        return values_list
+
+    def predict_batch(self, entity_list, top_n: int = 20, batch_size: int = 16):
+
+        if self._no_trainable_canonical_entity_map:
+            return [[] for _ in entity_list]
+
+        nbest_entities_list = []
+        results_list = []
+        for entity in entity_list:
+
+            if isinstance(entity, (list, tuple)):
+                top_entity = entity[0]
+                nbest_entities = tuple(entity)
             else:
-                self.fit()
+                top_entity = entity
+                nbest_entities = tuple([entity])
 
-        except EsConnectionError as e:
-            logger.error(
-                "Unable to connect to Elasticsearch: %s details: %s", e.error, e.info
+            nbest_entities_list.append(nbest_entities)
+
+            if self._is_system_entity:
+                # system entities are already resolved
+                results_list.append(top_entity.value)
+
+        if self._is_system_entity:
+            return results_list
+
+        results_list = self._predict_batch(nbest_entities_list, batch_size, top_n)
+
+        for i, results in enumerate(results_list):
+            if results:
+                results_list[i] = results[:top_n]
+
+        return results_list
+
+
+class TfIdfSparseCosSimEntityResolver(EntityResolverBase):
+    """
+    a tf-idf based entity resolver using sparse matrices. ref:
+    scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html
+    """
+
+    def __init__(self, app_path, entity_type, er_config, resource_loader, **_kwargs):
+        super().__init__(app_path, entity_type, resource_loader=resource_loader)
+
+        self._aug_lower_case = er_config.get("model_settings", {}).get("augment_lower_case", True)
+        self._aug_title_case = er_config.get("model_settings", {}).get("augment_title_case", False)
+        self._aug_normalized = er_config.get("model_settings", {}).get("augment_normalized", False)
+        self._aug_max_syn_embs = (
+            er_config.get("model_settings", {}).get("augment_max_synonyms_embeddings", True)
+        )
+
+        self._processed_entity_map = None
+        self._n = 5  # max number of character ngrams to consider
+        self._vectorizer = TfidfVectorizer(
+            analyzer=self._char_ngram_and_word_analyzer, lowercase=False)
+        self._syn_tfidf_matrix = None
+        self._unique_synonyms = []
+
+    def _char_ngram_and_word_analyzer(self, string):
+        results = []
+        # give more importance to starting and ending characters of a word
+        string = f" {string.strip()} "
+        for n in range(self._n + 1):
+            results.extend([''.join(gram) for gram in zip(*[string[i:] for i in range(n)])])
```

Comment:
There are also trailing/leading whitespaces in these char ngrams. Shouldnt they be removed?

Reply:
The trailing and leading spaces are kept to distinguish a character ngram between middle of a word or starting of a word. We use spaces instead of using something like sentence start and end tokens.